src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java

```diff
@@ -21,12 +26,22 @@
 @Service
 public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
+    private final CategoryRepository categoryRepository;
     private final SpecificationBuilder<Book> bookSpecificationBuilder;
     private final BookMapper bookMapper;
 
     @Override
     public BookDto save(CreateBookRequestDto bookRequestDto) {
         Book book = bookMapper.toBookEntity(bookRequestDto);
+
+        Set<Category> categories = bookRequestDto.getCategoryIds().stream()
+                .map(id -> categoryRepository.findById(id).orElseThrow(
```

Comment:
Also if you still want to validate categories it will be more efficient to grab all categories from db at once and check if the provided id exists in the obtained list. In the current implementation, for example, if book has 100 categories then 100 requests will be sent to db, such behavior not efficient